### DISCUSSION TITLE
Support git submodule updates by tags

### DISCUSSION THREAD

**Discussion by DanySK** [renovatebot/renovate#24827] [Suggest an Idea]
### What would you like Renovate to be able to do?

This feature request is the follow-up of [this discussion](https://github.com/renovatebot/renovate/discussions/14258).

Currently, renovate can update submodules, and does so following the specification in `.gitmodules`: if no branch is specified, it proposes updates pointing to the remote `HEAD`, otherwise, it proposes updates for the specified `branch`.

It would be nice to have a different strategy, based on tags, something like:
```
{
  "git-submodules": {
    "enabled": true,
    "prefer-follow-tags": true
  }
}
```
that works as follows:
> if a tag is checked out, and at least one tag is a valid semver version, then use git-refs and semver versioning and update using tags on the specified branch; else, update the branch as usual.

I'm not sure what the behaviour should be in case of tags with a higher version but located on a parent commit, or in case multiple versions are located at the same commit. We could provide options for both cases, though:

```
{
  "git-submodules": {
    "enabled": true,
    "prefer-follow-tags": true,
    "on-multiple-versions": "latest",
    "ignore-parent-commits": true
  }
}
```

possible alternatives for `on-multiple-versions`: `["latest", "oldest", "skip", "fail"]`

### If you have any ideas on how this should be implemented, please tell us here.

Requires git tree object inspection, which would be probably better realized by calling git. I'm tagging @rarkins and @mstoetzer who contributed to the initial discussion.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by viceice** [comment]
I don't like the config options, i would prefer to configure a Renovate comment in the submodules file.

That file is a simple ini file, so it should support comments.

So renovate can update the comment to a tag and do a digest update for the corresponding commit.

Digest pinning would be adding the renovate comment.

**Comment by DanySK** [comment]
OTOH, that would taint the configuration of git with information about renovate, and the option would not be shareable in a configuration.
Please note that the information in the `.gitmodules` file is not sufficient, as per [this discussion](https://github.com/renovatebot/renovate/discussions/14258): setting a tag there does not work as expected. Renovate apparently needs to fetch the current checked out revision by calling git.

**Comment by rarkins** [comment]
I also downvote this if it means adding new options like suggested. Instead, could we do this?

1. Use the branch name as `currentValue`
2. If that `currentValue` is a version, then look for newer "versions" in git refs (so either branch names or tag names would work)
3. If there's a newer version, it's treated like any other patch, minor or major update for Renovate
4. If you don't want this, then disable patch/minor/major updates for the `git-submodules` manager

**Comment by DanySK** [comment]
@rarkins sounds good, I propose this clarification:
* "branch name" is the name of the branch in `.gitmodules`
* "current commit" is the commit checked out

I tried to produce a diagram:
![activity diagram](http://www.plantuml.com/plantuml/svg/TPBBhjCm48RtUOfP9X67imS9jIALkcB52qoSSMBm9NZ6eBuzTjocQWGxsnb__yKviM2Kv6opBPh6Q6WRW1L5BU_E-13V7Vu84T8weH03UHmjJQ3pcIm4iv4tJgCBKxB4uD3ZJ34Z-lxyRUZxl58fHl82AZXd10p3SFb-_dgz372zd8SyqHMKX_P6tELjCO4WECE232FbJseXzJFBXYI7OO1FPP89ereW-5rkZEZLyhcEQEEdFGEwuEV2vp9VRHKTcuzD-HAAn06qSLeV3iV09F3s1IRIcEpZjBlBels3DwxkY_SgYjbB98V6W_6lPG0B2jNRB67D2x9CjP7AVfR2AocZ3Kq5yt4sSd-5GqC5K9pVswfDxMNag_acHD9S8rOLscnMln899QlGhuJMtluRhcOg3HvZLN8Dpnlq0pnrkFWr4Ipouo_fNYlmaz7DCqVT7vFe4F_rRks-fmL_cn0pkzd0fyn9plu5)

I think that we need to consider:
1. the information in `.gitmodules` to restrict the search of updates (all refs vs. current branch)
2. the information on the currently checked out commit to decide the type of updates (tag -> newer tags, branch -> same branch, random commit -> any ref)
3. the information from the configuration of the git-submodules manager to decide whether to default to the current behavior regardless all the previous stuff.

**Comment by rarkins** [comment]
I think that unfortunately this logic flow will not work as-is, because we do not pass rules about patch/minor/major into our extract function.

Can you help me understand what "a tag has been checked out" means? At the time of package extraction today, it's optional even init submodules. We generally only recommend initializing them prior to extraction unless you've got some package files referencing files *within a submodule* that would fail if the file is not found on the file system.

**Comment by DanySK** [comment]
OK, I had to study a little bit. The information we need are:
1. the information in .gitmodules to restrict the search of updates (all refs vs. current branch)
2. the currently checked out branch, which is memorized in the tree even if the submodule has not been checked out: `git ls-tree HEAD path/to/submodule` or `git submodule status`

In any case, we will need to fetch the submodule somehow, but that should something the system does already.

We can obtain all the needed information from `git submodule status`. Output examples:

```
 a845d40071d354de5c3efd39dbdf93fb73c0cab6 foo (heads/main)
+0cfab810f1804dd4e7d0d7a88767ff624d598bd5 bar (2.9.5)
-a845d40071d354de5c3efd39dbdf93fb73c0cab6 baz
```
which are, respectively: branch checked out, tag checked out, detached `HEAD`. The leading `+`, `U`, or `-` can be ignored, it identifies the status for initialized submodules.

So the whole set of information we can get is summarized in the following table, with the strategy I propose for extracting versions:

| submodule status / .gitmodules | no branch entry                      | branch = bar                                                            |
|------------------------------|--------------------------------------|-------------------------------------------------------------------------|
| branch: `hash (heads/foo)`   | fetch updates on refs/heads/foo      | fetch updates on refs/heads/bar                                         |
| tag: `hash (tag)`            | fetch updates from all upstream tags | fetch updates from upstream tags whose commits belong to refs/heads/bar |
| detached HEAD: `hash`        | current behavior                     | fetch updates on refs/heads/bar                                         |

What should the `extract` function return? The list of available versions?

**Comment by rarkins** [comment]
> In any case, we will need to fetch the submodule somehow, but that should something the system does already.

Can we avoid that during the extract phase? Because today we avoid cloning submodules in most cases and only need it if we are updating.

**Comment by rarkins** [comment]
Could you edit your examples above so that the table at the bottom (with three rows) maps to the three rows in `git submodule status`? I think I'll understand it easier that way.

Also can you clarify, does `git submodule status` provide all the info which `.gitmodules` does, or do we need to correlate both?

**Comment by mstoetzer** [comment]
> Also can you clarify, does git submodule status provide all the info which .gitmodules does, or do we need to correlate both?

Providing my .dot-files repo as an example to get the difference, how tags and main branches will look like. I have 3 submodules in use:

- `tools/nvm` uses the tag `v0.39.1` official nvm repo
- `zsh/oh-my-zsh` uses the latest commit from master/main branch
- `zsh/oh-my-zsh-custom/plugins/zsh-dircolors-solarized` uses the latest commit from master/main branch

`cat .gitmodules`

```
[submodule "tools/nvm"]
	path = tools/nvm
	url = https://github.com/nvm-sh/nvm.git
[submodule "zsh/oh-my-zsh"]
	path = zsh/oh-my-zsh
	url = https://github.com/ohmyzsh/ohmyzsh.git
[submodule "zsh/oh-my-zsh-custom/plugins/zsh-dircolors-solarized"]
	path = zsh/oh-my-zsh-custom/plugins/zsh-dircolors-solarized
	url = https://github.com/joel-porquet/zsh-dircolors-solarized.git
	shallow = true
```

`git submodule status`

```
9600617c52d0d2e48493424c529ac6c945d2775b tools/nvm (v0.39.1)
ef3f7c43a91eb2c90098843b0ee9193bb52cdc96 zsh/oh-my-zsh (remotes/origin/HEAD)
9c1c40e894130c2a4ece1fa30e95456f1750058b zsh/oh-my-zsh-custom/plugins/zsh-dircolors-solarized (heads/master)
``` 

Also to note here, if a tag exists for the latest commit `git submodule status` will display the tag. So we need a option to differentiate if the user wants renovate to update all commits from a specific branch or the tags only.

**Comment by rarkins** [comment]
Thanks. This is something I'm not familiar with. It's interesting to me that the nvm submodule in `.gitmodules` does not specify anything about `v0.39.1` yet that's what you have checked out. If you clone the same repo on a different computer/directory will you also get `v0.39.1`'s sha? I guess it must be stored somewhere inside `.git/` in that case

**Comment by mstoetzer** [comment]
@rarkins This provides a lot of information: https://stackoverflow.com/questions/5033441/where-does-git-store-the-sha1-of-the-commit-for-a-submodule :)

**Comment by DanySK** [comment]
@rarkins we do not need to *actually* check out the submodules, but we do need to find if new revisions exist in the remote repository. But this is something renovate does already somehow, I do not expect relevant changes to the core logic.

We indeed need to merge the information in `.gitmodules` with the information (from the tree) generated by `git submodule status`. `.gitmodules` says (optionally) which branch we are interested in, `git submodule status` reports what will be checked out when the submodule will get pulled. I think that this is the information that is currently changed by renovate when new updates are suggested.

**Comment by DanySK** [comment]
@mstoetzer waiting for @rarkins feedback: I won't begin implementing a feature until the maintainers confirm that it is design the way they want and that it will get merged.

**Comment by kramvan1** [comment]
+1 for the new feature.    Would also like to see the ability to skip pre-release tags, so only pull in the 
Can filter the tags with
`git tag -l | grep -E '^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$'` to skip the pre-release tags that are not `v.1.2.3`
or maybe there's a version mode to handle this already that can be applied to submodules somehow.

**Comment by pl4nty** [comment]
Renovate seems to detect new tags/digests with a versioning override and `branch = `, but it doesn't write the updated tag to `.gitmodules`. I wrote [a patch](https://github.com/pl4nty/renovate/tree/patch-1) to fix that.

I'm doing more testing though - I suspect the patch will write `branch = main` on digest updates if the branch isn't specified in `.gitmodules`. We probably want to avoid that in case the default branch changes, to match current behaviour.

```json
{
      "matchManagers": ["git-submodules"],
      "versioning": "semver-coerced"
}
```

**Reply by DanySK** [reply]
Has the patch been merged?

**Reply by Alexander-Shukaev** [reply]
Is this working already?

**Reply by Shegox** [reply]
I created #30104 inspired by this fork

**Comment by neumantm** [comment]
I've found the following workaround (it only cares about tags and completely ignores branches):

1. Disable the `git-submodules` module of renovatebot
2. Remove the `branch=` part in `.gitmodules`
3. Make sure the name of your submodules in `.gitmodules` (that is the part in quotation marks in the `[submodule ...]` line) is exactly the path of your repo within your git forge (e.g. if the project of the submodule is `github.com/example/foo` the name must be `example/foo`.
4. In your repository create a file called `.submoduleVersions` with the following structure:

```
<name of submodule1 from .gitmodules>=<gitTag for submodule1>@<gitHash>
<name of submodule2 from .gitmodules>=<gitTag for submodule2>@<gitHash>
...
```

5. Add the following to your renovate bot `config.js`:

```json
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["^\.submoduleVersions$"],
      "matchStrings": ["(?<depName>.*?)=(?<currentValue>.*?)@(?<currentDigest>.*?)\\n"],
      "packageNameTemplate": "https://github.com/{{{depName}}}",
      "datasourceTemplate": "git-refs",
    }
  ],
  "postUpgradeTasks": {
    "commands": ["cat .submoduleVersions | grep -E '^{{{depName}}}=' | sed -e 's/.*@//' | while read gitHash ; do submodulePath=\"$(git config --file .gitmodules submodule.\"{{{depName}}}\".path)\"; (cd \"$submodulePath\" ; git checkout $gitHash) ;done"],
    "executionMode": "update"
  },
```

and 

```json
  "allowedPostUpgradeCommands": [
    "cat submoduleVersions.ini | grep -E '.*' | sed -e 's/.*@//' | while read gitHash ; do submodulePath=\"$(git config --file \.gitmodules submodule\.\".*\"\.path)\"; (cd \"$submodulePath\" ; git checkout $gitHash) ;done"
  ],
  "allowPostUpgradeCommandTemplating": true,
```

**Reply by DanySK** [reply]
:exploding_head: :clap: :clap: :clap: :clap: :clap:
